**This is a movie recommender system in which we will create a model on the basis of some details of movies and it will recommend the movie that you will love to watch**

- There are different types of recommendee system
- This is a content based Movie recommender system
- On the basis of content you like, Director, Cast you like, it will recommend movies

In [2]:
import numpy as np
import pandas as pd

In [3]:
movies=pd.read_csv(r'tmdb_5000_movies.csv')
credits=pd.read_csv(r'tmdb_5000_credits.csv')

In [4]:
movies.head()

budget                                             genres  \
0  237000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
1  300000000  [{"id": 12, "name": "Adventure"}, {"id": 14, "...   
2  245000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
3  250000000  [{"id": 28, "name": "Action"}, {"id": 80, "nam...   
4  260000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   

                                       homepage      id  \
0                   http://www.avatarmovie.com/   19995   
1  http://disney.go.com/disneypictures/pirates/     285   
2   http://www.sonypictures.com/movies/spectre/  206647   
3            http://www.thedarkknightrises.com/   49026   
4          http://movies.disney.com/john-carter   49529   

                                            keywords original_language  \
0  [{"id": 1463, "name": "culture clash"}, {"id":...                en   
1  [{"id": 270, "name": "ocean"}, {"id": 726, "na...                en   
2  [{"id": 470, "name": "spy"}, {"id": 818, "name...                en   
3  [{"id": 849, "name": "dc comics"}, {"id": 853,...                en   
4  [{"id": 818, "name": "based on novel"}, {"id":...                en   

                             original_title  \
0                                    Avatar   
1  Pirates of the Caribbean: At World's End   
2                                   Spectre   
3                     The Dark Knight Rises   
4                               John Carter   

                                            overview  popularity  \
0  In the 22nd century, a paraplegic Marine is di...  150.437577   
1  Captain Barbossa, long believed to be dead, ha...  139.082615   
2  A cryptic message from Bond’s past sends him o...  107.376788   
3  Following the death of District Attorney Harve...  112.312950   
4  John Carter is a war-weary, former military ca...   43.926995   

                                production_companies  \
0  [{"name": "Ingenious Film Partners", "id": 289...   
1  [{"name": "Walt Disney Pictures", "id": 2}, {"...   
2  [{"name": "Columbia Pictures", "id": 5}, {"nam...   
3  [{"name": "Legendary Pictures", "id": 923}, {"...   
4        [{"name": "Walt Disney Pictures", "id": 2}]   

                                production_countries release_date     revenue  \
0  [{"iso_3166_1": "US", "name": "United States o...   2009-12-10  2787965087   
1  [{"iso_3166_1": "US", "name": "United States o...   2007-05-19   961000000   
2  [{"iso_3166_1": "GB", "name": "United Kingdom"...   2015-10-26   880674609   
3  [{"iso_3166_1": "US", "name": "United States o...   2012-07-16  1084939099   
4  [{"iso_3166_1": "US", "name": "United States o...   2012-03-07   284139100   

   runtime                                   spoken_languages    status  \
0    162.0  [{"iso_639_1": "en", "name": "English"}, {"iso...  Released   
1    169.0           [{"iso_639_1": "en", "name": "English"}]  Released   
2    148.0  [{"iso_639_1": "fr", "name": "Fran\u00e7ais"},...  Released   
3    165.0           [{"iso_639_1": "en", "name": "English"}]  Released   
4    132.0           [{"iso_639_1": "en", "name": "English"}]  Released   

                                          tagline  \
0                     Enter the World of Pandora.   
1  At the end of the world, the adventure begins.   
2                           A Plan No One Escapes   
3                                 The Legend Ends   
4            Lost in our world, found in another.   

                                      title  vote_average  vote_count  
0                                    Avatar           7.2       11800  
1  Pirates of the Caribbean: At World's End           6.9        4500  
2                                   Spectre           6.3        4466  
3                     The Dark Knight Rises           7.6        9106  
4                               John Carter           6.1        2124

In [5]:
movies.shape

(4803, 20)

In [6]:
credits.head()

movie_id                                     title  \
0     19995                                    Avatar   
1       285  Pirates of the Caribbean: At World's End   
2    206647                                   Spectre   
3     49026                     The Dark Knight Rises   
4     49529                               John Carter   

                                                cast  \
0  [{"cast_id": 242, "character": "Jake Sully", "...   
1  [{"cast_id": 4, "character": "Captain Jack Spa...   
2  [{"cast_id": 1, "character": "James Bond", "cr...   
3  [{"cast_id": 2, "character": "Bruce Wayne / Ba...   
4  [{"cast_id": 5, "character": "John Carter", "c...   

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...  
1  [{"credit_id": "52fe4232c3a36847f800b579", "de...  
2  [{"credit_id": "54805967c3a36829b5002c41", "de...  
3  [{"credit_id": "52fe4781c3a36847f81398c3", "de...  
4  [{"credit_id": "52fe479ac3a36847f813eaa3", "de...

In [7]:
credits.head(1)['crew']

0    [{"credit_id": "52fe48009251416c750aca23", "de...
Name: crew, dtype: object

In [8]:
credits.shape

(4803, 4)

# Now I will merge the 2 data frame on the basis of title column

In [10]:
movies.merge(credits,on='title').shape

(4809, 23)

**Now I reassign the data into a variable movies again**

In [12]:
movies=movies.merge(credits,on='title')

In [13]:
movies.head(1)

budget                                             genres  \
0  237000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   

                      homepage     id  \
0  http://www.avatarmovie.com/  19995   

                                            keywords original_language  \
0  [{"id": 1463, "name": "culture clash"}, {"id":...                en   

  original_title                                           overview  \
0         Avatar  In the 22nd century, a paraplegic Marine is di...   

   popularity                               production_companies  ... runtime  \
0  150.437577  [{"name": "Ingenious Film Partners", "id": 289...  ...   162.0   

                                    spoken_languages    status  \
0  [{"iso_639_1": "en", "name": "English"}, {"iso...  Released   

                       tagline   title vote_average vote_count movie_id  \
0  Enter the World of Pandora.  Avatar          7.2      11800    19995   

                                                cast  \
0  [{"cast_id": 242, "character": "Jake Sully", "...   

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...  

[1 rows x 23 columns]

# Now I will drop the unnecessary columns...

In [15]:
movies=movies[['movie_id','title','overview','genres','keywords','cast','crew']]

In [16]:
movies

movie_id                                     title  \
0        19995                                    Avatar   
1          285  Pirates of the Caribbean: At World's End   
2       206647                                   Spectre   
3        49026                     The Dark Knight Rises   
4        49529                               John Carter   
...        ...                                       ...   
4804      9367                               El Mariachi   
4805     72766                                 Newlyweds   
4806    231617                 Signed, Sealed, Delivered   
4807    126186                          Shanghai Calling   
4808     25975                         My Date with Drew   

                                               overview  \
0     In the 22nd century, a paraplegic Marine is di...   
1     Captain Barbossa, long believed to be dead, ha...   
2     A cryptic message from Bond’s past sends him o...   
3     Following the death of District Attorney Harve...   
4     John Carter is a war-weary, former military ca...   
...                                                 ...   
4804  El Mariachi just wants to play his guitar and ...   
4805  A newlywed couple's honeymoon is upended by th...   
4806  "Signed, Sealed, Delivered" introduces a dedic...   
4807  When ambitious New York attorney Sam is sent t...   
4808  Ever since the second grade when he first saw ...   

                                                 genres  \
0     [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
1     [{"id": 12, "name": "Adventure"}, {"id": 14, "...   
2     [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
3     [{"id": 28, "name": "Action"}, {"id": 80, "nam...   
4     [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
...                                                 ...   
4804  [{"id": 28, "name": "Action"}, {"id": 80, "nam...   
4805  [{"id": 35, "name": "Comedy"}, {"id": 10749, "...   
4806  [{"id": 35, "name": "Comedy"}, {"id": 18, "nam...   
4807                                                 []   
4808                [{"id": 99, "name": "Documentary"}]   

                                               keywords  \
0     [{"id": 1463, "name": "culture clash"}, {"id":...   
1     [{"id": 270, "name": "ocean"}, {"id": 726, "na...   
2     [{"id": 470, "name": "spy"}, {"id": 818, "name...   
3     [{"id": 849, "name": "dc comics"}, {"id": 853,...   
4     [{"id": 818, "name": "based on novel"}, {"id":...   
...                                                 ...   
4804  [{"id": 5616, "name": "united states\u2013mexi...   
4805                                                 []   
4806  [{"id": 248, "name": "date"}, {"id": 699, "nam...   
4807                                                 []   
4808  [{"id": 1523, "name": "obsession"}, {"id": 224...   

                                                   cast  \
0     [{"cast_id": 242, "character": "Jake Sully", "...   
1     [{"cast_id": 4, "character": "Captain Jack Spa...   
2     [{"cast_id": 1, "character": "James Bond", "cr...   
3     [{"cast_id": 2, "character": "Bruce Wayne / Ba...   
4     [{"cast_id": 5, "character": "John Carter", "c...   
...                                                 ...   
4804  [{"cast_id": 1, "character": "El Mariachi", "c...   
4805  [{"cast_id": 1, "character": "Buzzy", "credit_...   
4806  [{"cast_id": 8, "character": "Oliver O\u2019To...   
4807  [{"cast_id": 3, "character": "Sam", "credit_id...   
4808  [{"cast_id": 3, "character": "Herself", "credi...   

                                                   crew  
0     [{"credit_id": "52fe48009251416c750aca23", "de...  
1     [{"credit_id": "52fe4232c3a36847f800b579", "de...  
2     [{"credit_id": "54805967c3a36829b5002c41", "de...  
3     [{"credit_id": "52fe4781c3a36847f81398c3", "de...  
4     [{"credit_id": "52fe479ac3a36847f813eaa3", "de...  
...                                                 ...  
4804  [{"credit_id": "52fe44eec3a36847f80b2

# I will do preprocessing of the data

In [18]:
movies.isnull().sum()

movie_id    0
title       0
overview    3
genres      0
keywords    0
cast        0
crew        0
dtype: int64

In [19]:
from warnings import filterwarnings
filterwarnings('ignore')
movies.dropna(inplace=True)

In [20]:
movies.isnull().sum()

movie_id    0
title       0
overview    0
genres      0
keywords    0
cast        0
crew        0
dtype: int64

In [21]:
movies.duplicated().sum()

0

- Now I want to merge the columns like 'overview','genres','cast','crew'
- But you can see that the dat in those columns are very weiredly written
- So we have to preprocess it

In [23]:
movies.iloc[0].genres

'[{"id": 28, "name": "Action"}, {"id": 12, "name": "Adventure"}, {"id": 14, "name": "Fantasy"}, {"id": 878, "name": "Science Fiction"}]'

- We have to separate the 'Action','Adventure','Fantasy','SciFi' means the name part only
- Its looks like a list of dictionary but it is actually a string of list
- So 1st of all we need to convert it into a list
- For that we have a module in python as 'ast' and there is also a method called literal_eval()
- That will work to convert a string of list to list only
- We can see a example

In [25]:
import ast
ast.literal_eval('[{"id": 28, "name": "Action"}, {"id": 12, "name": "Adventure"}, {"id": 14, "name": "Fantasy"}, {"id": 878, "name": "Science Fiction"}]')

[{'id': 28, 'name': 'Action'},
 {'id': 12, 'name': 'Adventure'},
 {'id': 14, 'name': 'Fantasy'},
 {'id': 878, 'name': 'Science Fiction'}]

- And now we create function which will convert the data of 'genre' columns into that format
- we will see below

In [27]:
def convert(obj):
    list=[]
    for i in ast.literal_eval(obj):
        list.append(i['name'])
    return list

- And now we will apply this function into movies an then store it in that column only

In [29]:
movies['genres']=movies['genres'].apply(convert)

In [30]:
movies.head(1)

movie_id   title                                           overview  \
0     19995  Avatar  In the 22nd century, a paraplegic Marine is di...   

                                          genres  \
0  [Action, Adventure, Fantasy, Science Fiction]   

                                            keywords  \
0  [{"id": 1463, "name": "culture clash"}, {"id":...   

                                                cast  \
0  [{"cast_id": 242, "character": "Jake Sully", "...   

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...

# Now we will do the same with the column 'Keywords'

In [32]:
movies['keywords']=movies['keywords'].apply(convert)

In [33]:
movies.head()

movie_id                                     title  \
0     19995                                    Avatar   
1       285  Pirates of the Caribbean: At World's End   
2    206647                                   Spectre   
3     49026                     The Dark Knight Rises   
4     49529                               John Carter   

                                            overview  \
0  In the 22nd century, a paraplegic Marine is di...   
1  Captain Barbossa, long believed to be dead, ha...   
2  A cryptic message from Bond’s past sends him o...   
3  Following the death of District Attorney Harve...   
4  John Carter is a war-weary, former military ca...   

                                          genres  \
0  [Action, Adventure, Fantasy, Science Fiction]   
1                   [Adventure, Fantasy, Action]   
2                     [Action, Adventure, Crime]   
3               [Action, Crime, Drama, Thriller]   
4           [Action, Adventure, Science Fiction]   

                                            keywords  \
0  [culture clash, future, space war, space colon...   
1  [ocean, drug abuse, exotic island, east india ...   
2  [spy, based on novel, secret agent, sequel, mi...   
3  [dc comics, crime fighter, terrorist, secret i...   
4  [based on novel, mars, medallion, space travel...   

                                                cast  \
0  [{"cast_id": 242, "character": "Jake Sully", "...   
1  [{"cast_id": 4, "character": "Captain Jack Spa...   
2  [{"cast_id": 1, "character": "James Bond", "cr...   
3  [{"cast_id": 2, "character": "Bruce Wayne / Ba...   
4  [{"cast_id": 5, "character": "John Carter", "c...   

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...  
1  [{"credit_id": "52fe4232c3a36847f800b579", "de...  
2  [{"credit_id": "54805967c3a36829b5002c41", "de...  
3  [{"credit_id": "52fe4781c3a36847f81398c3", "de...  
4  [{"credit_id": "52fe479ac3a36847f813eaa3", "de...

- From the 'cast' column we need only 1st 3 dictionary because the movies are recommended on the basis of main characters 
- Thats why I think 1st 3 charcters are enough
- We will use the same function but we will add something because we only need 1st 3 names
- So here

In [35]:
def new(obj):
    list=[]
    count=0
    for i in ast.literal_eval(obj):
        if count<=2:
            list.append(i['name'])
            count+=1
        else:
            break
    return list

In [36]:
movies['cast']=movies['cast'].apply(new)

In [37]:
movies.head()

movie_id                                     title  \
0     19995                                    Avatar   
1       285  Pirates of the Caribbean: At World's End   
2    206647                                   Spectre   
3     49026                     The Dark Knight Rises   
4     49529                               John Carter   

                                            overview  \
0  In the 22nd century, a paraplegic Marine is di...   
1  Captain Barbossa, long believed to be dead, ha...   
2  A cryptic message from Bond’s past sends him o...   
3  Following the death of District Attorney Harve...   
4  John Carter is a war-weary, former military ca...   

                                          genres  \
0  [Action, Adventure, Fantasy, Science Fiction]   
1                   [Adventure, Fantasy, Action]   
2                     [Action, Adventure, Crime]   
3               [Action, Crime, Drama, Thriller]   
4           [Action, Adventure, Science Fiction]   

                                            keywords  \
0  [culture clash, future, space war, space colon...   
1  [ocean, drug abuse, exotic island, east india ...   
2  [spy, based on novel, secret agent, sequel, mi...   
3  [dc comics, crime fighter, terrorist, secret i...   
4  [based on novel, mars, medallion, space travel...   

                                               cast  \
0  [Sam Worthington, Zoe Saldana, Sigourney Weaver]   
1     [Johnny Depp, Orlando Bloom, Keira Knightley]   
2      [Daniel Craig, Christoph Waltz, Léa Seydoux]   
3      [Christian Bale, Michael Caine, Gary Oldman]   
4    [Taylor Kitsch, Lynn Collins, Samantha Morton]   

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...  
1  [{"credit_id": "52fe4232c3a36847f800b579", "de...  
2  [{"credit_id": "54805967c3a36829b5002c41", "de...  
3  [{"credit_id": "52fe4781c3a36847f81398c3", "de...  
4  [{"credit_id": "52fe479ac3a36847f813eaa3", "de...

- From the column 'crew' we only need the directors name
- we know very well now a days some people are watch movvies on the basis of director also
- we need to extract the directors name from crew
- if we see the crew data

In [39]:
movies['crew'][4808]

'[{"credit_id": "58ce021b9251415a390165d9", "department": "Production", "gender": 2, "id": 6888, "job": "Executive Producer", "name": "Clark Peterson"}, {"credit_id": "58ce0232c3a36850e90157da", "department": "Production", "gender": 2, "id": 61051, "job": "Executive Producer", "name": "Andrew Reimer"}, {"credit_id": "52fe44e8c3a368484e03da8d", "department": "Directing", "gender": 2, "id": 85563, "job": "Director", "name": "Brian Herzlinger"}, {"credit_id": "52fe44e8c3a368484e03da87", "department": "Directing", "gender": 2, "id": 94471, "job": "Director", "name": "Jon Gunn"}, {"credit_id": "52fe44e8c3a368484e03da97", "department": "Directing", "gender": 0, "id": 997560, "job": "Director", "name": "Brett Winn"}]'

- We need to extract where 'job' is Director and the name of that director 
- For tht we need to modify the function again

In [41]:
def cre(obj):
    list=[]
    for i in ast.literal_eval(obj):
        if i['job']=="Director":
            list.append(i['name'])
            break
    return list

# Now we will apply the function on 'crew' column

In [43]:
movies['crew']=movies['crew'].apply(cre)

In [44]:
movies.head()

movie_id                                     title  \
0     19995                                    Avatar   
1       285  Pirates of the Caribbean: At World's End   
2    206647                                   Spectre   
3     49026                     The Dark Knight Rises   
4     49529                               John Carter   

                                            overview  \
0  In the 22nd century, a paraplegic Marine is di...   
1  Captain Barbossa, long believed to be dead, ha...   
2  A cryptic message from Bond’s past sends him o...   
3  Following the death of District Attorney Harve...   
4  John Carter is a war-weary, former military ca...   

                                          genres  \
0  [Action, Adventure, Fantasy, Science Fiction]   
1                   [Adventure, Fantasy, Action]   
2                     [Action, Adventure, Crime]   
3               [Action, Crime, Drama, Thriller]   
4           [Action, Adventure, Science Fiction]   

                                            keywords  \
0  [culture clash, future, space war, space colon...   
1  [ocean, drug abuse, exotic island, east india ...   
2  [spy, based on novel, secret agent, sequel, mi...   
3  [dc comics, crime fighter, terrorist, secret i...   
4  [based on novel, mars, medallion, space travel...   

                                               cast                 crew  
0  [Sam Worthington, Zoe Saldana, Sigourney Weaver]      [James Cameron]  
1     [Johnny Depp, Orlando Bloom, Keira Knightley]     [Gore Verbinski]  
2      [Daniel Craig, Christoph Waltz, Léa Seydoux]         [Sam Mendes]  
3      [Christian Bale, Michael Caine, Gary Oldman]  [Christopher Nolan]  
4    [Taylor Kitsch, Lynn Collins, Samantha Morton]     [Andrew Stanton]

# Now we can see in the column 'overview' all data are in string format

In [46]:
movies['overview'][0]

'In the 22nd century, a paraplegic Marine is dispatched to the moon Pandora on a unique mission, but becomes torn between following orders and protecting an alien civilization.'

- So I have a plan to concatenate it with other columns thats why I have to convert it into a list
- So by using lambda function I convert it into a list

In [48]:
movies['overview']=movies['overview'].apply(lambda x:x.split())

In [49]:
movies.head()

movie_id                                     title  \
0     19995                                    Avatar   
1       285  Pirates of the Caribbean: At World's End   
2    206647                                   Spectre   
3     49026                     The Dark Knight Rises   
4     49529                               John Carter   

                                            overview  \
0  [In, the, 22nd, century,, a, paraplegic, Marin...   
1  [Captain, Barbossa,, long, believed, to, be, d...   
2  [A, cryptic, message, from, Bond’s, past, send...   
3  [Following, the, death, of, District, Attorney...   
4  [John, Carter, is, a, war-weary,, former, mili...   

                                          genres  \
0  [Action, Adventure, Fantasy, Science Fiction]   
1                   [Adventure, Fantasy, Action]   
2                     [Action, Adventure, Crime]   
3               [Action, Crime, Drama, Thriller]   
4           [Action, Adventure, Science Fiction]   

                                            keywords  \
0  [culture clash, future, space war, space colon...   
1  [ocean, drug abuse, exotic island, east india ...   
2  [spy, based on novel, secret agent, sequel, mi...   
3  [dc comics, crime fighter, terrorist, secret i...   
4  [based on novel, mars, medallion, space travel...   

                                               cast                 crew  
0  [Sam Worthington, Zoe Saldana, Sigourney Weaver]      [James Cameron]  
1     [Johnny Depp, Orlando Bloom, Keira Knightley]     [Gore Verbinski]  
2      [Daniel Craig, Christoph Waltz, Léa Seydoux]         [Sam Mendes]  
3      [Christian Bale, Michael Caine, Gary Oldman]  [Christopher Nolan]  
4    [Taylor Kitsch, Lynn Collins, Samantha Morton]     [Andrew Stanton]

- I already mentioned that we will go to merge some columns 
- So for that we have to remove the spaces from the words from some columns
- Because, For example in column 'cast' ther is a name called 'Sam Worthington', and in the 
     - column 'crew' there is also a name called 'Sam Mendes'
- And if we did'nt remove the space from those word then in future the machine can divide the name 'Sam Worthington' into 'Sam' and 'Worthington'
- And 'Sam Mendes' into 'Sam' and 'Mendes'
- So by that our model can face the problem in future 
- Because you search for the movie by 'Sam Mendes' and it will recommend you a movie of 'Sam Worthington' 
- Because of the same tag 'Sam'
- Thats why we have to remove the space from the name and make it one word only
- We will do the same work in all columns and make those words into one word

In [51]:
movies['cast']=movies['cast'].apply(lambda x:[i.replace(" ","") for i in x])
movies['crew']=movies['crew'].apply(lambda x:[i.replace(" ","") for i in x])
movies['keywords']=movies['keywords'].apply(lambda x:[i.replace(" ","") for i in x])
movies['genres']=movies['genres'].apply(lambda x:[i.replace(" ","") for i in x])

In [52]:
movies.head()

movie_id                                     title  \
0     19995                                    Avatar   
1       285  Pirates of the Caribbean: At World's End   
2    206647                                   Spectre   
3     49026                     The Dark Knight Rises   
4     49529                               John Carter   

                                            overview  \
0  [In, the, 22nd, century,, a, paraplegic, Marin...   
1  [Captain, Barbossa,, long, believed, to, be, d...   
2  [A, cryptic, message, from, Bond’s, past, send...   
3  [Following, the, death, of, District, Attorney...   
4  [John, Carter, is, a, war-weary,, former, mili...   

                                         genres  \
0  [Action, Adventure, Fantasy, ScienceFiction]   
1                  [Adventure, Fantasy, Action]   
2                    [Action, Adventure, Crime]   
3              [Action, Crime, Drama, Thriller]   
4           [Action, Adventure, ScienceFiction]   

                                            keywords  \
0  [cultureclash, future, spacewar, spacecolony, ...   
1  [ocean, drugabuse, exoticisland, eastindiatrad...   
2  [spy, basedonnovel, secretagent, sequel, mi6, ...   
3  [dccomics, crimefighter, terrorist, secretiden...   
4  [basedonnovel, mars, medallion, spacetravel, p...   

                                            cast                crew  
0  [SamWorthington, ZoeSaldana, SigourneyWeaver]      [JamesCameron]  
1     [JohnnyDepp, OrlandoBloom, KeiraKnightley]     [GoreVerbinski]  
2      [DanielCraig, ChristophWaltz, LéaSeydoux]         [SamMendes]  
3      [ChristianBale, MichaelCaine, GaryOldman]  [ChristopherNolan]  
4    [TaylorKitsch, LynnCollins, SamanthaMorton]     [AndrewStanton]

- Now we gonna add a new columns named 'tags' by concatenating the columns like overview, genres, cast, keywords, crew

In [54]:
movies['tags']=movies['overview'] + movies['genres'] + movies['keywords'] + movies['cast'] + movies['crew']

In [55]:
movies.head()

movie_id                                     title  \
0     19995                                    Avatar   
1       285  Pirates of the Caribbean: At World's End   
2    206647                                   Spectre   
3     49026                     The Dark Knight Rises   
4     49529                               John Carter   

                                            overview  \
0  [In, the, 22nd, century,, a, paraplegic, Marin...   
1  [Captain, Barbossa,, long, believed, to, be, d...   
2  [A, cryptic, message, from, Bond’s, past, send...   
3  [Following, the, death, of, District, Attorney...   
4  [John, Carter, is, a, war-weary,, former, mili...   

                                         genres  \
0  [Action, Adventure, Fantasy, ScienceFiction]   
1                  [Adventure, Fantasy, Action]   
2                    [Action, Adventure, Crime]   
3              [Action, Crime, Drama, Thriller]   
4           [Action, Adventure, ScienceFiction]   

                                            keywords  \
0  [cultureclash, future, spacewar, spacecolony, ...   
1  [ocean, drugabuse, exoticisland, eastindiatrad...   
2  [spy, basedonnovel, secretagent, sequel, mi6, ...   
3  [dccomics, crimefighter, terrorist, secretiden...   
4  [basedonnovel, mars, medallion, spacetravel, p...   

                                            cast                crew  \
0  [SamWorthington, ZoeSaldana, SigourneyWeaver]      [JamesCameron]   
1     [JohnnyDepp, OrlandoBloom, KeiraKnightley]     [GoreVerbinski]   
2      [DanielCraig, ChristophWaltz, LéaSeydoux]         [SamMendes]   
3      [ChristianBale, MichaelCaine, GaryOldman]  [ChristopherNolan]   
4    [TaylorKitsch, LynnCollins, SamanthaMorton]     [AndrewStanton]   

                                                tags  
0  [In, the, 22nd, century,, a, paraplegic, Marin...  
1  [Captain, Barbossa,, long, believed, to, be, d...  
2  [A, cryptic, message, from, Bond’s, past, send...  
3  [Following, the, death, of, District, Attorney...  
4  [John, Carter, is, a, war-weary,, former, mili...

- Now we will drop the columns that we concatenate the 4 columns
- we will create a new Dataframe

In [57]:
recommend=movies[['movie_id','title','tags']]

In [58]:
recommend

movie_id                                     title  \
0        19995                                    Avatar   
1          285  Pirates of the Caribbean: At World's End   
2       206647                                   Spectre   
3        49026                     The Dark Knight Rises   
4        49529                               John Carter   
...        ...                                       ...   
4804      9367                               El Mariachi   
4805     72766                                 Newlyweds   
4806    231617                 Signed, Sealed, Delivered   
4807    126186                          Shanghai Calling   
4808     25975                         My Date with Drew   

                                                   tags  
0     [In, the, 22nd, century,, a, paraplegic, Marin...  
1     [Captain, Barbossa,, long, believed, to, be, d...  
2     [A, cryptic, message, from, Bond’s, past, send...  
3     [Following, the, death, of, District, Attorney...  
4     [John, Carter, is, a, war-weary,, former, mili...  
...                                                 ...  
4804  [El, Mariachi, just, wants, to, play, his, gui...  
4805  [A, newlywed, couple's, honeymoon, is, upended...  
4806  ["Signed,, Sealed,, Delivered", introduces, a,...  
4807  [When, ambitious, New, York, attorney, Sam, is...  
4808  [Ever, since, the, second, grade, when, he, fi...  

[4806 rows x 3 columns]

# Now I will convert the list into string

In [60]:
recommend['tags']=recommend['tags'].apply(lambda x:" ".join(x))

In [61]:
recommend

movie_id                                     title  \
0        19995                                    Avatar   
1          285  Pirates of the Caribbean: At World's End   
2       206647                                   Spectre   
3        49026                     The Dark Knight Rises   
4        49529                               John Carter   
...        ...                                       ...   
4804      9367                               El Mariachi   
4805     72766                                 Newlyweds   
4806    231617                 Signed, Sealed, Delivered   
4807    126186                          Shanghai Calling   
4808     25975                         My Date with Drew   

                                                   tags  
0     In the 22nd century, a paraplegic Marine is di...  
1     Captain Barbossa, long believed to be dead, ha...  
2     A cryptic message from Bond’s past sends him o...  
3     Following the death of District Attorney Harve...  
4     John Carter is a war-weary, former military ca...  
...                                                 ...  
4804  El Mariachi just wants to play his guitar and ...  
4805  A newlywed couple's honeymoon is upended by th...  
4806  "Signed, Sealed, Delivered" introduces a dedic...  
4807  When ambitious New York attorney Sam is sent t...  
4808  Ever since the second grade when he first saw ...  

[4806 rows x 3 columns]

# Convert all the data in tags into lower case

In [63]:
recommend['tags']=recommend['tags'].apply(lambda x:x.lower())

In [64]:
recommend

movie_id                                     title  \
0        19995                                    Avatar   
1          285  Pirates of the Caribbean: At World's End   
2       206647                                   Spectre   
3        49026                     The Dark Knight Rises   
4        49529                               John Carter   
...        ...                                       ...   
4804      9367                               El Mariachi   
4805     72766                                 Newlyweds   
4806    231617                 Signed, Sealed, Delivered   
4807    126186                          Shanghai Calling   
4808     25975                         My Date with Drew   

                                                   tags  
0     in the 22nd century, a paraplegic marine is di...  
1     captain barbossa, long believed to be dead, ha...  
2     a cryptic message from bond’s past sends him o...  
3     following the death of district attorney harve...  
4     john carter is a war-weary, former military ca...  
...                                                 ...  
4804  el mariachi just wants to play his guitar and ...  
4805  a newlywed couple's honeymoon is upended by th...  
4806  "signed, sealed, delivered" introduces a dedic...  
4807  when ambitious new york attorney sam is sent t...  
4808  ever since the second grade when he first saw ...  

[4806 rows x 3 columns]

In [65]:
recommend['tags'][0]

'in the 22nd century, a paraplegic marine is dispatched to the moon pandora on a unique mission, but becomes torn between following orders and protecting an alien civilization. action adventure fantasy sciencefiction cultureclash future spacewar spacecolony society spacetravel futuristic romance space alien tribe alienplanet cgi marine soldier battle loveaffair antiwar powerrelations mindandsoul 3d samworthington zoesaldana sigourneyweaver jamescameron'

- Now we will do vectorization
- When we will do vectorization we will remove stopwords

In [67]:
from sklearn.feature_extraction.text import CountVectorizer
counter=CountVectorizer(max_features=5000,stop_words='english')

In [68]:
vectors=counter.fit_transform(recommend['tags']).toarray()

In [69]:
vectors

array([[0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       ...,
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0]], dtype=int64)

In [70]:
vectors[0]

array([0, 0, 0, ..., 0, 0, 0], dtype=int64)

In [71]:
counter.get_feature_names_out()

array(['000', '007', '10', ..., 'zone', 'zoo', 'zooeydeschanel'],
      dtype=object)

- Now we will apply Stemming by which we will use one word for 3 other words of same root word
- Example: there are words like [love,loved,loving]
- After stemming we get only love (the root word) for those 3 words

In [73]:
import nltk
from nltk.stem.porter import PorterStemmer
ps=PorterStemmer()

- now we create a function by which we stem the 'tag' columns for one by one words
- and then we will make the list into string again

In [75]:
def stem(text):
    l=[]
    for i in text.split():
        l.append(ps.stem(i))
    return " ".join(l)

In [76]:
stem('in the 22nd century, a paraplegic marine is dispatched to the moon pandora on a unique mission, but becomes torn between following orders and protecting an alien civilization. action adventure fantasy sciencefiction cultureclash future spacewar spacecolony society spacetravel futuristic romance space alien tribe alienplanet cgi marine soldier battle loveaffair antiwar powerrelations mindandsoul 3d samworthington zoesaldana sigourneyweaver jamescameron')

'in the 22nd century, a parapleg marin is dispatch to the moon pandora on a uniqu mission, but becom torn between follow order and protect an alien civilization. action adventur fantasi sciencefict cultureclash futur spacewar spacecoloni societi spacetravel futurist romanc space alien tribe alienplanet cgi marin soldier battl loveaffair antiwar powerrel mindandsoul 3d samworthington zoesaldana sigourneyweav jamescameron'

- Now we will apply the same function on the recommend['tags']

In [78]:
recommend['tags']=recommend['tags'].apply(stem)

- Here we have 4806 movies and for each there are 5000 words
- Now we are talking about a 5000 dimension plane
- And now we have to calculate the distance from each other of movies
- If the distance is low then those movies are of same type
- And if the distance is high then the movies are of different genre
- We are going to calculate the Cosine distance instead of Eucledian distance
- The Eucledian distance calculate from tip to tip
- And the Cosine distance calculate the angle between 2 vectors
- When you work on a high dimenssional data, the Eucledian distance fails there, Its not reliable there


In [80]:
from sklearn.metrics.pairwise import cosine_similarity

In [81]:
distance=cosine_similarity(vectors)

In [82]:
distance[0]

array([1.        , 0.08964215, 0.05976143, ..., 0.02519763, 0.02817181,
       0.        ])

- Now we will create a function and provide a movie name
- Then the function will fetch the index of that movie from our data
- And the it will find the similarity or distance
- From our distance variable where all the distance from one movie to another is stored
- Then it will sort the distance 
- By that the lower distance from given movie comes first and going further
- But there is also a problem arise
- When we will sort the distance then the index will change 
- And it will make trouble for us
- Thats why we will use 'enumerate', by which it will change into a list of tuple
- Then we will fetch the top 5 similar movies of the given movie

In [84]:
def find(movie_name):
    movie_index=recommend[recommend['title']==movie_name].index[0]
    similar_movie=distance[movie_index]
    movie_list=sorted(list(enumerate(similar_movie)),reverse=True,key=lambda x:x[1])[1:6]
    # This line is for sorting the similar movies on the basis of similarities without changing index number and show the data after 1st movie, 
    # Because the given movie and the 1st movie similarity is always 1
    for i in movie_list:
        print(recommend.iloc[i[0]].title)

In [85]:
find('Batman Begins')

The Dark Knight
The Dark Knight Rises
Batman
Batman & Robin
Batman


- Now we will save the model in a pickle file

In [87]:
import pickle
pickle.dump(recommend,open('Movies_recommender.pkl','wb'))

In [88]:
pickle.dump(distance,open('Distance.pkl','wb'))